### PR TITLE
[core] fixed CONNECTION_STATE_MCS_CREATE_REQUEST parsing

### PR DESCRIPTION
**Pull request by akallabeth**
* Fix various issues with state machine
* Fix some memory leaks
* Improve logging
* Add missing server side parsing code

### CODE REVIEW

File: libfreerdp/core/nla.c [hunk 1]
```diff
@@ -1453,20 +1462,23 @@ static int nla_decode_ts_request(rdpNla* nla, wStream* s)
 		return -1;
 	}
```
Comment (hardening) [inline libfreerdp/core/nla.c:1464]: From my PoV changes before are ok, but the next changes that save the result of decoding in `rc` are useless and make the more complex to read.

Comment (akallabeth) [reply libfreerdp/core/nla.c:1464]: didn´t like the handling of the return as boolean value (`!value`) so I used the `rc=fkt(); if (rc == 0) return;` style

File: libfreerdp/core/nla.c [hunk 2]
```diff
@@ -1432,13 +1433,21 @@ static int nla_decode_ts_request(rdpNla* nla, wStream* s)
 	WLog_DBG(TAG, "<<----- receiving...");
 
 	/* TSRequest */
-	if (!WinPrAsn1DecReadSequence(&dec, &dec2))
+	const size_t offset = WinPrAsn1DecReadSequence(&dec, &dec2);
+	if (offset == 0)
+		return -1;
+	if (Stream_GetRemainingLength(s) < offset)
```
[the remark this answers is not in the document]
Comment (akallabeth) [reply libfreerdp/core/nla.c]: removed

**Review by mfleisz — COMMENTED** [review]
Overall LGTM

I agree with @hardening - the changes in `nla.c` bloats the code, makes it harder to read (best example is the `while` below) and introduces a mixed API usage style compared to how the ASN API is used i.e. in the SSPI kerberos/negotiate implementations.

File: libfreerdp/core/nla.c (hunk 1 above)
Comment (akallabeth) [reply libfreerdp/core/nla.c:1462]: ok, changed to use similar style as before but check for `xxx() == 0` instead of using a return variable.

**Comment by akallabeth** [comment]
@hardening @mfleisz ok, update the pr accordingly.

**Review by hardening — APPROVED** [review]

**Pull request merged**